onnxruntime/core/session/custom_ops.cc

```diff
@@ -395,17 +398,272 @@ struct CustomOpKernel : OpKernel {
+    } else {
+      def_builder.TypeConstraint(input_name, DataTypeImpl::TensorTypeFromONNXEnum(static_cast<int>(input_type))->AsTensorType());
+    }
+  }
+
+  for (size_t i = 0; i < output_count; i++) {
+    const auto output_type = op->GetOutputType(op, i);
+    const auto output_name = "Output" + std::to_string(i);
+    if (output_type == ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED) {
+      def_builder.TypeConstraint(output_name, DataTypeImpl::AllTensorTypes());
+    } else {
+      def_builder.TypeConstraint(output_name, DataTypeImpl::TensorTypeFromONNXEnum(static_cast<int>(output_type))->AsTensorType());
+    }
+  }
+
+  if (const char* provider_type = op->GetExecutionProviderType(op)) {
+    def_builder.Provider(provider_type);
+  } else {
+    def_builder.Provider(onnxruntime::kCpuExecutionProvider);
+  }
+
+  KernelCreateFn kernel_create_fn = [op](FuncManager&, const OpKernelInfo& info, std::unique_ptr<OpKernel>& out) -> Status {
+    out = std::make_unique<CustomOpKernel>(info, *op);
+    return Status::OK();
+  };
+
+  return KernelCreateInfo(def_builder.Build(), kernel_create_fn);
+}
+
+ONNX_NAMESPACE::OpSchema CreateSchema(const std::string& domain, const OrtCustomOp* op) {
+  constexpr uint32_t min_ort_version_with_optional_io_support = 8;
+  constexpr uint32_t min_ort_version_with_variadic_io_support = 14;
+
+  const size_t input_count = op->GetInputTypeCount(op);
+  const size_t output_count = op->GetOutputTypeCount(op);
+  int undefined = 0;
+
+  ONNX_NAMESPACE::OpSchema schema(op->GetName(op), "custom op registered at runtime", 0);
+
+  for (size_t i = 0; i < input_count; i++) {
+    onnx::OpSchema::FormalParameterOption option = onnx::OpSchema::FormalParameterOption::Single;
+    bool is_homogeneous = true;
+    int min_arity = 1;
+
+    // The OrtCustomOp interface did not support the methods to query input/output characteristics before
+    // ORT API version 8. So, query the relevant methods ONLY from API version 8 onwards.
+    if (op->version >= min_ort_version_with_optional_io_support) {
+      const auto characteristic = op->GetInputCharacteristic(op, i);
+
+      // Support for optional and variadic inputs/output was added in versions 8 and 14, respectively.
+      if (characteristic == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_OPTIONAL) {
+        option = onnx::OpSchema::FormalParameterOption::Optional;
+      } else if ((op->version >= min_ort_version_with_variadic_io_support) &&
+                 (characteristic == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_VARIADIC)) {
+        ORT_ENFORCE(i == input_count - 1, "Only the last input to a custom op may be marked variadic.");
+        option = onnx::OpSchema::FormalParameterOption::Variadic;
+        min_arity = op->GetVariadicInputMinArity(op);
+        is_homogeneous = static_cast<bool>(op->GetVariadicInputHomogeneity(op));
+      }
+    }
+
+    const auto type = op->GetInputType(op, i);
```

Comment:
put some comments about the design here: make the type constrains support all the types, and handle the type inference in TypeShapeInference func.

Reply:
Done.